vertical-pod-autoscaler/pkg/utils/metrics/quality/quality.go

```diff
@@ -67,3 +112,59 @@ func ObserveUsageRecommendationRelativeDiff(usage, recommendation float64, isOOM
 func ObserveMissingRecommendation(isOOM bool, resource corev1.ResourceName) {
 	usageMissingRecommendationCounter.WithLabelValues(string(resource), strconv.FormatBool(isOOM)).Inc()
 }
+
+// ObserveUsageRecommendationDiff records absolute diff between usage and
+// recommendation.
+func ObserveUsageRecommendationDiff(usage, recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	switch resource {
+	case corev1.ResourceCPU:
+		cpuUsageRecommendationDiff.Observe((usage - recommendation))
+	case corev1.ResourceMemory:
+		memoryUsageRecommendationDiff.WithLabelValues(strconv.FormatBool(isOOM)).Observe(usage - recommendation)
+	default:
+		klog.Warningf("Unknown resource: %v", resource)
+	}
+}
+
+// ObserveRecommendation records the value of recommendation.
+func ObserveRecommendation(recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	switch resource {
+	case corev1.ResourceCPU:
+		cpuRecommendations.Observe(recommendation)
+	case corev1.ResourceMemory:
+		memoryRecommendations.WithLabelValues(strconv.FormatBool(isOOM)).Observe(recommendation)
+	default:
+		klog.Warningf("Unknown resource: %v", resource)
+	}
+}
+
+// ObserveQualityMetrics records all quality metrics that we can derive from usage and recommendation.
+func ObserveQualityMetrics(usage, recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	ObserveRecommendation(recommendation, isOOM, resource)
+	ObserveUsageRecommendationDiff(usage, recommendation, isOOM, resource)
+	ObserveUsageRecommendationRelativeDiff(usage, recommendation, isOOM, resource)
+}
+
+// ObserveQualityMetricsRecommendationMissing records all quality metrics that we can derive from usage when recommendation is missing.
+func ObserveQualityMetricsRecommendationMissing(usage float64, isOOM bool, resource corev1.ResourceName) {
+	ObserveMissingRecommendation(isOOM, resource)
+	ObserveUsageRecommendationDiff(usage, 0, isOOM, resource)
```

Comment:
I think it would be useful to label these entries with "recommendation missing".

Reply:
Good point, done,